kable-btleplug-ffi/src/peripheral.rs

```diff
@@ -0,0 +1,257 @@
+                let adapter = get_adapter().await;
+                let mut events = adapter.events().await.unwrap();
+                loop {
+                    tokio::select! {
+                        _ = token.cancelled() => {
+                            break;
+                        },
+                        Some(event) = events.next() => match event {
+                            CentralEvent::DeviceConnected(connected) =>
+                                if connected == id.platform { callbacks.connected() }
+                            CentralEvent::DeviceDisconnected(disconnected) =>
+                                if disconnected == id.platform { callbacks.disconnected() }
+                            _ => {}
+                        },
+                    }
+                }
+            });
+        });
+
+        peripheral
+    }
+
+    async fn properties(&self) -> Result<PeripheralProperties> {
+        let platform = self.get_platform().await?.properties().await?.unwrap();
+        Ok(PeripheralProperties::new(self.id.clone(), platform))
+    }
+
+    async fn connect(&self, cancellation_handle: Arc<CancellationHandle>) -> bool {
+        if let Ok(platform) = self.get_platform().await {
+            return self
+                .platform_connect(platform, cancellation_handle)
+                .await
+                .is_ok();
+        }
+
+        // Ooph. We're in a bad place because btleplug removes peripherals from the adapter when
+        // they disconnect. This means that we have to re-scan until the device shows back up in
+        // the adapter and then attempt connection.
+        let adapter = get_adapter().await;
+        let mut events = match adapter.events().await {
+            Ok(events) => events,
+            Err(_) => return false,
+        };
+        if let Err(_) = adapter.start_scan(ScanFilter::default()).await {
+            return false;
+        }
+
+        let peripheral_token = self.cancellation.token();
+        let connect_token = cancellation_handle.token();
+        loop {
+            tokio::select! {
+                _ = peripheral_token.cancelled() => break,
+                _ = connect_token.cancelled() => break,
+                Some(event) = events.next() => match event {
+                    CentralEvent::DeviceConnected(_) | CentralEvent::DeviceUpdated(_) => {
+                        if adapter.peripheral(&self.id.platform).await.is_ok() { break; }
+                    },
+                    _ => {}
+                }
+            }
+        }
+
+        if let Ok(platform) = self.get_platform().await {
+            self.platform_connect(platform, cancellation_handle)
+                .await
+                .is_ok()
+        } else {
+            false
+        }
+    }
+
+    async fn disconnect(&self) -> bool {
+        match self.get_platform().await {
+            Err(_) => true,
+            Ok(platform) => platform.disconnect().await.is_ok(),
```

Comment:
This is probably fine, but is confusing to me. Can you please explain why we return `true` on error, and `true` on Ok, and `false` on "not Ok"?

Reply:
Cool, thanks for explaining.